OTA: rework uploader; extend log; increase timeouts

RubenKelevra opened the pull request:
# What does this implement/fix?

* Nagle is a no-op on large transfers. It merges small writes, which we do not do, so disabling it slightly trims the tail delay.
* Remove the artificial send buffer limit (SO\_SNDBUF). No longer needed since we do not use a blocking socket.
* Send the image as one continuous stream so TCP can keep more data in flight and recover faster from packet loss. Avoid periodic pauses every 8 kByte.
* Clamp TCP MSS to 536 so more packets fit into the microcontroller's small receive window. This increases the chance of fast recovery on loss without changing the microcontroller RWND. Overhead impact is minimal.
* Use a non-blocking socket and process microcontroller responses asynchronously. Progress is now driven primarily by TCP ACKs observed via SIOCOUTQ on Linux; when this is not available, fall back to the microcontroller's 8 kByte CHUNK\_OK acknowledgements.
* Progress bar: starts at 0, shows percentages with 2 decimals. If TCP ACKs can be read, shows "Segments: a/b" to reflect the microcontroller ACK bytes. Updates are throttled and only forced if there are not enough socket events to keep output responsive.
* Extend and clarify setup logging. Status messages clarify what is happening before the firmware image transfer starts and indicate where it might hang.
* Improve error handling: detect stalls (no send progress or protocol ACKs within the transfer timeout), report early peer close, and flag unexpected bytes during upload.
* Increase transfer timeout to 600 s for lossy or spotty links.
* Increase connect timeout to 60 s. This does not mean we actually wait that long; it is a safeguard that gives TCP enough leeway to handle ACK resends and decide when to abort.

## Test results:


### Speed to an ESP32-S3 with good connectivity

TL;DR: New Updater and TCP SACK is faster than the old method without TCP SACK. The smaller packages don't have any impact on the performance.

#### With the old updater:

```
INFO Upload took 6.76 seconds, waiting for result...
```

#### With sack and the old method:

```
INFO Upload took 4.42 seconds, waiting for result...
```

#### With the new updater, without sack:

```
INFO Upload took 23.46 seconds, waiting for result...
```


#### With the new updater and sack:

```
INFO Upload took 4.58 seconds, waiting for result...
```


### Speed to an ESP32-C3 Supermini with bad connectivity:

TL;DR: The old method didn't work, if timeouts were massively increased, it would take 15 minutes for an update. New method with SACK is the fasted option and consistently up to ~13 times faster.

#### Before any change:

- Timeout at various stages, even when run every half hour for 5 days no luck doing an update

#### With the old updater but longer timeouts and smaller TCP packages:

```
INFO Upload took 898.68 seconds, waiting for result...
```

#### With the new updater, without sack - good example:

```
INFO Upload took 164.96 seconds, waiting for result...
```

#### With the new updater, without sack - bad example:

(Sometimes it would go *extreeeemly* slow, much slower than that example below. I never completed these runs.)
```
INFO Upload took 254.26 seconds, waiting for result...
```

#### With the new updater and sack:

Consistently good results:

```
INFO Upload took 82.81 seconds, waiting for result...
--- 
INFO Upload took 71.05 seconds, waiting for result...
---
INFO Upload took 112.56 seconds, waiting for result...
```

Note:

Sometimes it *will* still be slower in the worst case scenarios, but I had a deep look at the traffic, and I could not see a single stall. The sender tried to recover efficiently and did pick up as soon as the RF conditions improved again.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

The issue:

https://github.com/esphome/esphome/pull/10152#issuecomment-3173587115

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32 (ESP32-C3 Supermini)
- [x] ESP32 IDF (ESP32-S3 and ESP32-C3 Supermini - with and without TCP SACK)
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840




## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

bdraco commented on esphome/components/esp32/__init__.py:
```diff
@@ -912,6 +912,8 @@ async def to_code(config):
         if advanced.get(CONF_ENABLE_LWIP_CHECK_THREAD_SAFETY, True):
             add_idf_sdkconfig_option("CONFIG_LWIP_CHECK_THREAD_SAFETY", True)
 
+        add_idf_sdkconfig_option("CONFIG_LWIP_TCP_SACK_OUT", True)
```
This is only going to work on ESP-IDF for ESP32

bdraco replied:
I'd split this into a separate PR so we can focus on testing the client changes with and without it

RubenKelevra replied:
As I said: The Linux kernel does not support TCP without SACK well. It's pretty legacy to even use TCP without SACK, so I'm not surprised that nobody noticed.

Enabling SACK is a complete no-brainer, if 1 KByte of additional flash usage is not an issue.

RubenKelevra replied:
Noted that in the PR and the commit message. All other TCP SACK options will be send as individual PRs. But since I've tested TCP SACK for ESP32 / ESP-IDF extensively and it improves the situation for all network conditions, while also being the standard option for all modern OS, it's safe to merge like this I think.

RubenKelevra commented on esphome/components/esp32/__init__.py:
```diff
@@ -970,6 +972,8 @@ async def to_code(config):
         cg.add_build_flag("-DUSE_ESP32_FRAMEWORK_ARDUINO")
         cg.add_platformio_option("platform_packages", [conf[CONF_SOURCE]])
 
+        cg.add_define("CONFIG_LWIP_TCP_SACK_OUT", 1)
```
No idea if that will work. Haven't tested that. I'm too tired.

RubenKelevra commented:
Agreed, that would be a good idea. Go for it. I also don't have the hardware - as I still don't think this will eventually take off ;)

Wasn't there something about supporting Zigbee too? Or do I remember that wrong? How's OTA going over Zigbee? Does this use the Zigbee build in OTA function then?

kepstin commented:
So, first, a caveat: While these tests were done in quick succession, there is no guarantee that the network conditions are directly comparable. There are other (Matter) devices live on the network as the OTA tests are being performed, causing unpredictable traffic.

#### ESPHome 2025.7.5 (baseline, without SACK)

```
INFO Uploading /home/cwalton/Development/esphome/.esphome/build/fancontroller/.pioenvs/fancontroller/firmware.bin (903648 bytes)
INFO Upload took 363.78 seconds, waiting for result...
```

Speed: 2484 byte/s

#### ESPHome 2025.7.5 (with SACK)

```
INFO Uploading /home/cwalton/Development/esphome/.esphome/build/fancontroller/.pioenvs/fancontroller/firmware.bin (903648 bytes)
INFO Upload took 279.06 seconds, waiting for result...
```

Speed: 3237 byte/s

#### New uploader (without SACK)

```
INFO Uploading /home/cwalton/Development/esphome/.esphome/build/fancontroller/.pioenvs/fancontroller/firmware.bin (903520 bytes)
INFO Upload took 673.83 seconds, waiting for result...

```

Speed: 1341 byte/s (there were several stalls during this upload)

#### New uploader (with SACK)

```
INFO Uploading /home/cwalton/Development/esphome/.esphome/build/fancontroller/.pioenvs/fancontroller/firmware.bin (903520 bytes)
INFO Upload took 352.95 seconds, waiting for result...

```

Speed: 2560 byte/s

---

So, I'm not sure how valid these conclusions are given the lack of control over the network environment, but here's what I see over Thread:

* SACK helps quite a bit, and should be enabled even if the other uploader changes are not accepted.
* The old updater seems to be less likely to stall when SACK isn't enabled.
* The old updater seems to be able to achieve higher throughput than the new uploader when SACK is enabled (I wonder if the higher MSS helps, somehow?).

RubenKelevra commented:
> Zigbee doesn't use TCP/IP (I honestly don't know how they're planning to do OTA? Probably use zigbee native functionality.), so I don't think this change would be applicable.

Yeah I know, that Zigbee itself doesn't support TCP/IP. But you would need to provide the image to the Zigbee "hub" somehow.

I'm using z2m, and this is just fetching Zigbee updates via `http(s)://`. So ESPHome could either provide the image via n URL, or maybe just push "through" z2m directly to the device instead? That was what I was thinking about.

z2m or ZHA does the translation from ESPHome's push to Zigbee, and we get the feedback back how far we're already pushed and if the update is successful etc.

Sorry for the tangent. I was just wondering if there's already something in place I'm not yet familiar with.

RubenKelevra commented:
Thanks a lot for your test!


> So, I'm not sure how valid these conclusions are given the lack of control over the network environment, but here's what I see over Thread:
> 
>     * SACK helps quite a bit, and should be enabled even if the other uploader changes are not accepted.

Yeah, that's not surprising. Can you tell me what host OS did the push?



> 
>     * The old updater seems to be less likely to stall when SACK isn't enabled.

That conclusion is off. The longer duration you’re seeing comes from **smaller TCP segments** after I enabled **MSS clamping** (down to 536), not from the “new vs old” updater logic itself. Smaller segments do add some overhead, but they also give us more frequent feedback and quicker recovery when there’s loss. That trade-off is intentional.

This matters differently depending on the transport:

**Wi-Fi (802.11n or higher)**  

Wi-Fi can aggregate small packets (A-MSDU / A-MPDU), so the extra overhead mostly lands in TCP bookkeeping. In practice it’s tiny and rarely noticeable.

**Thread (802.15.4 + 6LoWPAN)**  

There’s no link-layer aggregation. Even though IPv6 MTU is 1280 bytes, anything that size is sent as *many 6LoWPAN fragments* (typically ~80–100 bytes payload each).  

- The “old” updater sent ~1280-byte packets → lots of fragments.  
- The “new” updater (MSS 536) sends smaller packets → fewer fragments per packet.

Yes, that means more headers overall, but it also means each transmission occupies the channel for less time. In a mesh, that’s healthier: lower burstiness, more chances for other nodes to get airtime, and usually better latency under load.


>     * The old updater seems to be able to achieve higher throughput than the new uploader when SACK is enabled (I wonder if the higher MSS helps, somehow? Higher MSS means that the lower Thread layer is doing more of the work segmenting and re-assembling the IP packets. I think Thread might be able to retransmit individual  sub-128-byte segments of the IP packet, possibly with low latency on a hop-by-hop basis, during this re-assembly process.).

You’re right about the segmentation/reassembly mechanics — Thread can retransmit individual small fragments hop-by-hop (IIRC). But the idea that “bigger MSS = better” doesn’t really hold up in a mesh.

The new updater forces a smaller MSS (536 bytes), which adds some header overhead but shortens each transmission burst. That means less channel hogging, more room for other nodes to get a word in, and typically lower latency under load.

So yes, it may shave a bit of peak throughput in a perfect link, but at the cost of latency of other data transfer and with much less feedback if there's loss. In the messy real world, it's a net win for the whole network to use smaller packages.


> 
> It would be neat to have a "dry run" firmware update mode for testing this where the device accepts and verifies the upload, but then throws away the data without actually _applying_ the update.

The flash of my devices would agree. If you want to avoid just the update and reboot, you can send a wrong MD5 sum, and it should reject it.

But yeah, throwing the data away and instead just running a fuzzy latency loop instead of a flash operation would be nice for the flash ... or in case of PSRAM devices, it could be written to the PSRAM and the MD5 sum checked.

Not sure however if that's worth the effort, as it needs client support.

RubenKelevra commented:
I've removed the not yet working ESP32 Arduino framework setting, and added the detail to the commit message and the PR description. Also fixed the bug where the updater would announce that it got the OTA version, while this step was only send, but not read back.

Should be fine to merge from my perspective. But I would appreciate more testing, if someone can spare some time! :)

RubenKelevra commented:
> > Yeah, that's not surprising. Can you tell me what host OS did the push?
> 
> The host os was a Linux system (kernel 6.14), running over wifi 6 to the ap, then ethernet to the thread border router.

Yeah that explains why the improvement using SACK is so big. I'll have a look at the Kernelaand the RFCs for TCP to figure out if my suspicion is correct. While SACK will always work better, we won't be able to use it on all types of devices. On ESP8266 for example seems to be no support available, despite documentation that there is. 



> > Smaller segments do add some overhead, but they also give us more frequent feedback and quicker recovery when there’s loss.
> 
> Unfortunately, at such low transfer rates, the progress bar does stop moving for long periods of the time, giving the _impression_ of stalls even if that's not what's really happening.

Yep, but this feedback is completely decoupled from the MSS clamping. We just send currently a single byte resonse every 8 KByte, which stems from the old update method of sending out chunks. 

We could basically switch that to 1 KByte ACKs to get more responsiveness out of the progress bar. 

Alternatively we could tap deeper in the socket to read the ACK themselves 😬

Haven't yet looked at the API on python for that, though.

RubenKelevra commented:
> > > Smaller segments do add some overhead, but they also give us more frequent feedback and quicker recovery when there’s loss.
> > 
> > 
> > Unfortunately, at such low transfer rates, the progress bar does stop moving for long periods of the time, giving the _impression_ of stalls even if that's not what's really happening.
> 
> Yep, but this feedback is completely decoupled from the MSS clamping. We just send currently a single byte resonse every 8 KByte, which stems from the old update method of sending out chunks.
> 
> We could basically switch that to 1 KByte ACKs to get more responsiveness out of the progress bar.
> 
> Alternatively we could tap deeper in the socket to read the ACK themselves 😬
> 
> Haven't yet looked at the API on python for that, though.

@kepstin
That's actually dead simple, but the API is platform dependent.

I will just add a packet based progress bar for Linux and the rest falls back to the 8 KByte blocks. I think that makes sense based on effort and user basis. If someone wants to implement that for MacOS or similar, that should be easy as well, just need to ask a slightly different API for the same thing.

bdraco reviewed (changes requested):
Thanks for this PR! The OTA improvements look great, especially the switch to non-blocking sockets and continuous streaming.

I wanted to highlight one important consideration: The TCP SACK (Selective Acknowledgment) feature you're enabling with `add_idf_sdkconfig_option("CONFIG_LWIP_TCP_SACK_OUT", True)` will only work for ESP-IDF builds on ESP32. 

For Arduino framework builds on ANY platform (ESP32, ESP8266, RP2040, LibreTiny), this won't have any effect because:
- All Arduino cores come with pre-compiled libraries including their network stacks
- The compile-time defines are already baked into the Arduino framework binaries
- ESP8266 Arduino uses lwIP pre-compiled without SACK support
- LibreTiny platforms have their own pre-compiled network implementations

This means that SACK benefits will only apply to ESP32 devices using ESP-IDF. All Arduino framework users across all platforms won't get this particular enhancement. However, they'll still benefit from all the other improvements you've made (smaller packets, non-blocking sockets, better timeouts, etc.).

**The SACK enablement needs to be moved to a separate PR** and must follow the same pattern as the other advanced lwIP options in the ESP32 component. It should be implemented as a configurable option under `esp32: → framework: → advanced:` like the existing `enable_lwip_*` options (e.g., `enable_lwip_tcp_sack: true` with a default of `true`). This maintains consistency with the existing lwIP configuration pattern and allows us to revert just the SACK change if issues arise, without losing all the OTA improvements.

The rest of the changes look solid and should improve OTA reliability across both frameworks!

The PR currently has several code quality issues that need to be addressed before it can be merged:

1. **Code style violations** - The Python code in `espota2.py` needs to pass the project's linting requirements (ruff, flake8)

2. **Duplicate code in logging functions** - The `log_info_inline`, `log_inline_append`, and `log_newline` functions all duplicate the pattern of getting stream handlers and writing to them. This should be refactored to reduce duplication.

3. **Function complexity** - The new `image_transfer_with_device_acks_nonblocking()` function is too long and complex. Specifically, the main `while acked < total:` loop with the selector logic should be extracted into its own function for better readability and maintainability.

4. **Missing docstrings** - New functions should have proper docstrings explaining their purpose and parameters

5. **Line length** - Some lines may exceed the project's line length limits

6. **Overly broad exception handling** - The code uses `with contextlib.suppress(Exception):` which catches all exceptions. This should be narrowed to only the specific exceptions that are expected (e.g., `AttributeError`, `ValueError`, etc.)

7. **Unnecessary try/except for constants** - This code pattern is nonsensical:
   ```python
   try:
       ack_block_size = OTA_IMAGE_ACK_SIZE_OTA_V2
   except NameError as e:
       raise ValueError("Global OTA_IMAGE_ACK_SIZE_OTA_V2 is not defined...") from e
   ```
   `OTA_IMAGE_ACK_SIZE_OTA_V2` is a constant defined in the same file. If it's not defined, the module wouldn't even import. This try/except should be removed.

To fix these, I suggest:
- Extract the common stream handler writing logic into a shared helper function
- Move the `while acked < total:` selector loop into a separate function like `_transfer_with_acks()` or similar
- Replace broad `Exception` catches with specific exception types
- Remove the unnecessary try/except blocks around constant assignments
- Run `pre-commit run --all-files` locally to catch all linting errors
- Add comprehensive docstrings to new functions
- Consider further breaking down complex logic into smaller, focused functions

Once these code quality issues are resolved, the functional improvements look great!

RubenKelevra committed:
OTA: refactor uploader; TCP ACK progress; bump timeouts

- Nagle is a no-op on large transfers. It merges small writes, which we do
  not do, so disabling it slightly trims the tail delay.
- Remove the artificial send buffer limit (SO_SNDBUF). No longer needed
  since we do not use a blocking socket.
- Send the image as one continuous stream so TCP can keep more data in
  flight and recover faster from packet loss. Avoid periodic pauses every
  8 kByte.
- Clamp TCP MSS to 536 so more packets fit into the microcontroller's small
  receive window. This increases the chance of fast recovery on loss
  without changing the microcontroller RWND. Overhead impact is minimal.
- Use a non-blocking socket and process microcontroller responses
  asynchronously. Progress is now driven primarily by TCP ACKs observed via
  SIOCOUTQ on Linux; when this is not available, fall back to the
  microcontroller's 8 kByte CHUNK_OK acknowledgements.
- Progress bar: starts at 0, shows percentages with 2 decimals. If TCP ACKs
  can be read, shows "Segments: a/b" to reflect the microcontroller ACK
  bytes. Updates are throttled and only forced if there are not enough
  socket events to keep output responsive.
- Extend and clarify setup logging. Status messages clarify what is
  happening before the firmware image transfer starts and indicate where it
  might hang.
- Improve error handling: detect stalls (no send progress or protocol ACKs
  within the transfer timeout), report early peer close, and flag unexpected
  bytes during upload.
- Increase transfer timeout to 600 s for lossy or spotty links.
- Increase connect timeout to 60 s. This does not mean we actually wait
  that long; it is a safeguard that gives TCP enough leeway to handle ACK
  resends and decide when to abort.

RubenKelevra commented:
@kepstin please retest, the progress bar should now update at least every 300 ms, if there's any progress received by TCP ACKs.

The ACK bytes, which drove the progress bar before, are now shown as "Segments: a/b", if the new TCP ACK based progress is active.

If the progress bar does not show "Segments: A/B" behind the percent, the TCP ACK feature is not working (works only under Linux).

Hope that improves the situation for you.

@bdraco did the rewrite you've asked me to do. Will have another fresh look tomorrow, if I really addressed every concern in your review and then mark it as "ready for review".

bdraco commented:
Thank you for the OTA refactor work. The improvements to the upload process are valuable additions. However, there are some remaining architectural concerns that should be addressed before merging:

---

`pylint` is failing
```
************ Module esphome.espota2
esphome/espota2.py:542: [R1720(no-else-raise), perform_ota] Unnecessary "else" after "raise", remove the "else" and de-indent the code inside it
esphome/espota2.py:739: [R1705(no-else-return), _transfer_with_acks._render_progress] Unnecessary "else" after "return", remove the "else" and de-indent the code inside it
esphome/espota2.py:748: [W0212(protected-access), _transfer_with_acks._render_progress] Access to a protected member _last_percent_int of a client class
esphome/espota2.py:749: [W0212(protected-access), _transfer_with_acks._render_progress] Access to a protected member _last_segments of a client class
esphome/espota2.py:762: [W0212(protected-access), _transfer_with_acks._render_progress] Access to a protected member _last_percent_int of a client class
esphome/espota2.py:763: [W0212(protected-access), _transfer_with_acks._render_progress] Access to a protected member _last_segments of a client class
esphome/espota2.py:777: [W0212(protected-access), _transfer_with_acks._render_progress] Access to a protected member _last_percent_int of a client class
esphome/espota2.py:778: [W0212(protected-access), _transfer_with_acks._render_progress] Access to a protected member _last_segments of a client class
```


---

`_transfer_with_acks` Function Complexity:

The `_transfer_with_acks` function has become too complex with nested functions and excessive use of `nonlocal` variables. The `_render_progress()` nested function with multiple `nonlocal` declarations is a clear indication this should be refactored into a class:

```python
def _render_progress() -> bool:
    nonlocal acked_tcp, use_sock_ack, acked_fallback  # Code smell
```

Create an `OTATransferHandler` class to encapsulate this state and logic properly.

---

Use `struct.pack` for Binary Encoding:

The `_encode_upload_size` function manually constructs bytes. This should use Python's `struct` module for clarity and consistency:


---

Extract `perform_ota` Try Block:

The `perform_ota` function has a massive try block (lines 434-571) that should be extracted into a separate function for better readability:

```python
def perform_ota(remote_host: str, remote_port: int, password: str, filename: str) -> None:
    """Send the OTA image to the remote host."""
    sock = _connect_to_device(remote_host, remote_port)
    try:
        _execute_ota_upload(sock, password, filename)
    finally:
        sock.close()

def _connect_to_device(remote_host: str, remote_port: int) -> socket.socket:
    """Establish connection to the OTA device."""
    # Connection logic here
    
def _execute_ota_upload(sock: socket.socket, password: str, filename: str) -> None:
    """Execute the actual OTA upload protocol."""
    # All the current try block content here
```

something like....

```python
class OTATransferHandler:
    """Handles OTA firmware transfer with progress tracking."""
    
    def __init__(self, sock: socket.socket, data: bytes, 
                 ack_block_size: int, progress: ProgressBar | None = None):
        self.sock = sock
        self.data = memoryview(data)
        self.total = len(data)
        self.ack_block_size = ack_block_size
        self.progress = progress
        
        # Transfer state
        self.sent = 0
        self.acked_device = 0
        self.acked_segments = 0
        self.acked_tcp = 0
        self.use_sock_ack = False
        
        # Timing
        self.last_activity = time.perf_counter()
        self.last_update = self.last_activity
        
    def transfer(self) -> None:
        """Execute the transfer with ACK handling."""
        # Main transfer logic here
        
    def _update_progress(self) -> bool:
        """Update progress bar if needed."""
        # Progress rendering logic here
        
    def _handle_socket_write(self) -> int:
        """Handle socket write events."""
        # Write handling logic here
        
    def _handle_device_acks(self, data: bytes) -> None:
        """Process device acknowledgments."""
        # ACK processing logic here
```

nits:

The progress bar throttling logic could be simplified with a dedicated timer class
Error messages could be more descriptive about what the user should do to resolve issues

**This PR needs testing on Windows** to ensure:
- The fallback mechanism works correctly when `fcntl` is unavailable
- Progress reporting functions properly without TCP ACK tracking
- No import errors or exceptions occur on non-Linux platforms
- The OTA upload completes successfully despite the platform differences

Please consider these refactoring suggestions to improve code maintainability before merging. The functionality is solid, but the code structure needs improvement to meet the project's quality standards.

nielsnl68 commented:
Ping me when it is far enough, and i will be happy to test this on my windows system.

RubenKelevra commented:
@nielsnl68 please do, the feedback would be valuable now. I did as much testing as I could under Linux. Meaning: The fallback works fine, if I "break" the check for the Linux platform. Question is more if I overlooked something.

@bdraco:

The pylint check was what I was referring to, with another pass. Will look at the other points, tomorrow, too! :)

RubenKelevra committed:
address pylint concerns

kepstin commented:
The current version of this PR has introduced a regression (might have been present previously, not sure), where it no longer resolves IPv6 addresses (Thread is IPv6-only, so there is no IPv4 address):

dev branch with this change applied:
```
INFO Successfully compiled program.
INFO Uploading to fancontroller.local
INFO Resolving IP address of fancontroller.local in mDNS
ERROR Could not resolve IPv4 address for fancontroller.local
```

If I try specifying the IPv6 address instead of the mDNS hostname, it still fails:
```
INFO Successfully compiled program.
INFO Uploading to fda9:d862:9078:1:61ef:eac8:1eec:49b6
ERROR Could not resolve IPv4 address for fda9:d862:9078:1:61ef:eac8:1eec:49b6
```

For reference, here's the output of the dev branch without this change applied:
```
INFO Successfully compiled program.
INFO Uploading to fancontroller.local
INFO Resolving IP address of fancontroller.local in mDNS
INFO Connecting to fda9:d862:9078:1:61ef:eac8:1eec:49b6 port 3232...
INFO Connected to fda9:d862:9078:1:61ef:eac8:1eec:49b6
INFO Uploading /home/cwalton/Development/esphome/.esphome/build/fancontroller/.pioenvs/fancontroller/firmware.bin (916336 bytes)
Uploading: [============================================================] 100% Done...

INFO Upload took 292.00 seconds, waiting for result...
INFO OTA successful
```

RubenKelevra commented:
@kepstin ah damn, didn't think about that. Thanks for the quick check!

RubenKelevra committed:
address ipv6 regression

RubenKelevra commented:
@kepstin please try again.

RubenKelevra committed:
ruff reformat

kepstin commented:
Thanks for the IPv6 fix. Everything looks like it's working as it should be - the percentage indicator is now updating approximately 8 times more often than the segments counter during the transfer.

There might be an off-by-one error in the segments counter? when the transfer was complete I see:
```
Uploading: [============================================================] 100.00 % - Segments: 111/112 Done...
```

When testing without SACK, i fairly often see updates to the percentage indicator pause for about 3 seconds at a time. the longest pause I saw was about 7 seconds. Testing with SACK, the percentage indicator updates were a lot more consistent - I never saw it pause for more than about 1 second at a time.

RubenKelevra commented:
> When testing without SACK, i fairly often see updates to the percentage indicator pause for about 3 seconds at a time. the longest pause I saw was about 7 seconds.

Yep, that's a Linux bug, I'm looking into next. Thanks for confirming :)

The display in this case is absolutely correct: Linux just "stalls" and won't send any new packages for quite a long time, because it falls back onto RTO backoff, instead of doing a fast retransmission, which it should.


> There might be an off-by-one error in the segments counter? when the transfer was complete I see:
> 
> ```
> Uploading: [============================================================] 100.00 % - Segments: 111/112 Done...
> ```

Weird. I look into that. Looks fine here, tho:

```
INFO Asking node for OTA version... OK
INFO Asking node for features... OK
INFO Authorize with node to do the update... OK
INFO Asking node if firmware image size can be accepted... OK
INFO Providing node with md5 checksum for firmware... OK
INFO Starting to send firmware image to node...
Uploading: [============================================================] 100.00 % - Segments: 155/155 Done...
```

> the percentage indicator is now updating approximately 8 times more often than the segments counter during the transfer.

Yeah, the granularity is actually even higher, so if the transfer speed is slower, you should see up to 15 updates before the segment counter goes up.

RubenKelevra commented:
> There might be an off-by-one error in the segments counter? when the transfer was complete I see:
> 
> ```
> Uploading: [============================================================] 100.00 % - Segments: 111/112 Done...
> ```

Found the bug, it's basically just a display error: If the socket is quicker in returning the "all packages acked" than the response Byte for the last segment, the progress bar will be "finished" without counting up the last segment.

I've added a fix in my refactoring.

jfroy commented:
Another vote to finish this PR. OTA over Thread is not reliable (i.e. never finishes due to timeouts) for many of my devices. I'm happy to help with testing.

RubenKelevra commented:
I'm sorry for the delay, some projects took all spare time I usually invest in OS work from me. But I finished the last one yesterday :)

Will finish this ASAP

jfroy commented:
Would you be open to someone else taking this across the finish line? I have so many thread devices I’d like to update without having to plug them in a computer 😅

bdraco commented:
> Would you be open to someone else taking this across the finish line? I have so many thread devices I’d like to update without having to plug them in a computer 😅

It's been inactive for several months at this point, so I think it's fair game. Feel free to pick it up!
